Distinguish between Unschedulable and UnschedulableAndUnresolvable in scheduler's PostFilter

kerthcet opened the pull request:
Before, in RunPostFilterPlugins, we didn't distinguish between unschedulable and unresolvable because we only have one postFilterPlugin by default, now, we have at least two, we should make sure that once a postFilterPlugin returns unresolvable, we'll return directly

Signed-off-by: Kante Yin <kerthcet@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
/sig scheduling

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/113854

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
When any scheduler plugin returns an `unschedulableAndUnresolvable` status
in `PostFilter`, the scheduling cycle terminates immediately for that Pod.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

kerthcet commented:
/ok-to-test

kerthcet commented:
/retest

ahg-g commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -766,8 +766,10 @@ func (f *frameworkImpl) RunPostFilterPlugins(ctx context.Context, state *framewo
+			// Any status other than Success, Unschedulable or UnschedulableAndUnresolvable is Error.
 			return nil, framework.AsStatus(s.AsError())
+		} else if s.IsUnresolvable() {
+			return r, s
```
Merge already handles this; there is no need for special handling unless you are planning to remove `PluginToStatus` completely (and if you do, lets do it as part of https://github.com/kubernetes/kubernetes/pull/114629)

kerthcet replied:
Yes, if any postFilter plugin returns `unschedulableAndUnresolvable`, we'll exit immediately.

sftim commented:
Suggestion for the changelog text:
```markdown
Changed the scheduler so that when any scheduler plugin returns an `unschedulableAndUnresolvable` status
in `PostFilter`, the scheduling cycle terminates immediately for that Pod.
```

If this isn't technically accurate, please revise the suggestion - I guessed based on the existing text, and haven't looked at the code changes.

kerthcet commented:
> ```gfm
> Changed the scheduler so that when any scheduler plugin returns an `unschedulableAndUnresolvable` status
> in `PostFilter`, the scheduling cycle terminates immediately for that Pod.
> ```

Thanks, updated.

Huang-Wei commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -751,7 +751,7 @@ func (f *frameworkImpl) runFilterPlugin(ctx context.Context, pl framework.Filter
 
 // RunPostFilterPlugins runs the set of configured PostFilter plugins until the first
-// Success or Error is met, otherwise continues to execute all plugins.
+// Success or Error or UnschedulableAndUnresolvable is met, otherwise continues to execute all plugins.
```
```suggestion
// Success, Error or UnschedulableAndUnresolvable is met; otherwise continues to execute all plugins.
```

Huang-Wei commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -184,7 +184,7 @@ func (ev *Evaluator) Preempt(ctx context.Context, pod *v1.Pod, m framework.NodeT
 		}
 		// Specify nominatedNodeName to clear the pod's nominatedNodeName status, if applicable.
-		return framework.NewPostFilterResultWithNominatedNode(""), framework.NewStatus(framework.Unschedulable, fitError.Error())
+		return framework.NewPostFilterResultWithNominatedNode(""), framework.NewStatus(framework.UnschedulableAndUnresolvable, fitError.Error())
```
This doesn't look right to me. The default preemption plugin is implemented in a **single-node** preemption manner, in other words, it hasn't iterated all preemption strategies. So we should keep `Unschedulable` to give other out-of-tree PostFilter plugins a chance to proceed with their custom logic (maybe cross-node preemption).

ahg-g replied:
+1; lets not change the current behavior of the existing preemption plugin.

kerthcet replied:
`We may have more preemption strategies` is right, revert the change to default preemption plugin.

Huang-Wei commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -766,8 +766,10 @@ func (f *frameworkImpl) RunPostFilterPlugins(ctx context.Context, state *framewo
-			// Any status other than Success or Unschedulable is Error.
+			// Any status other than Success, Unschedulable or UnschedulableAndUnresolvable is Error.
 			return nil, framework.AsStatus(s.AsError())
+		} else if s.IsUnresolvable() {
```
nit: merge with `if s.IsSuccess()`

ahg-g reviewed (commented):
We should remove PluginToStatus in this PR.

kerthcet commented on pkg/scheduler/framework/interface.go:
```diff
@@ -231,6 +232,11 @@ func (s *Status) IsUnschedulable() bool {
+// IsUnresolvable returns true if "Status" is UnschedulableAndUnresolvable.
+func (s *Status) IsUnresolvable() bool {
+	return s.Code() == UnschedulableAndUnresolvable
+}
```
It makes the code more lean.

ahg-g replied:
but the interface should be minimal, this is redundant.

kerthcet replied:
Removed.

kerthcet commented:
> We should remove PluginToStatus in this PR.

https://github.com/kubernetes/kubernetes/pull/114699/commits/975dd314d34dfd3f58771f21a6f870f877b01cec

kerthcet commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -752,30 +752,29 @@ func (f *frameworkImpl) runFilterPlugin(ctx context.Context, pl framework.Filter
 			result = r
 		}
-		statuses[pl.Name()] = s
+		status = status.Merge(s)
```
Another approach is only appending reasons here, which seems reasonable if we always return unschedulable status without error.

ahg-g replied:
That is what I had in mind as well, I prefer not to expose functions unless it is needed... and status.Merge is not needed now, so I think we should remove all of that code.

Huang-Wei replied:
> Another approach is only appending reasons here, which seems reasonable if we always return unschedulable status without error.

+1. And this is the only place `status.Merge()` is called, so we can remove it as well as `statusPrecedence`.

kerthcet replied:
Updated.

kerthcet commented:
/retest

ahg-g commented:
Thanks Kante, please squash.

kerthcet committed:
Distinguish unschedulable with unresolvable in scheduler

Before, in RunPostFilterPlugins, we didn't distinguish between unschedulable and unresolvable
because we only have one postFilterPlugin by default, now, we have at least two, we should
make sure that once a postFilterPlugin returns unresolvable, we'll return directly

Signed-off-by: Kante Yin <kerthcet@gmail.com>

kerthcet commented:
> Thanks Kante, please squash.

Squashed.

kerthcet commented:
/retest

Huang-Wei commented:
/lgtm
/approve

/retitle 
Distinguish between Unschedulable and UnschedulableAndUnresolvable in scheduler's PostFilter

ahg-g commented:
/lgtm
